netbox/utilities/jobs.py

```diff
@@ -0,0 +1,196 @@
+
+    The purpose of this class is to decouple jobs from the usual request-based approach. A practical example of this is
+    to schedule a periodic synchronization job for a particular object. All that matters is that the job is scheduled
+    and executed periodically. However, a new periodic job does not need to be scheduled every time the object is saved.
+    Calling the `schedule()` method of this class will ensure that the job's schedule is set up no matter how often the
+    method is called.
+    """
+
+    ENQUEUED_STATUS = [
+        JobStatusChoices.STATUS_PENDING,
+        JobStatusChoices.STATUS_SCHEDULED,
+        JobStatusChoices.STATUS_RUNNING,
+    ]
+
+    @classmethod
+    def get_jobs(cls, instance):
+        """
+        Get all jobs of this schedule related to a specific instance.
+        """
+        object_type = ObjectType.objects.get_for_model(instance, for_concrete_model=False)
+        return Job.objects.filter(
+            object_type=object_type,
+            object_id=instance.pk,
+            name=cls.__name__,
+        )
+
+    @classmethod
+    def enqueue(cls, *args, **kwargs):
+        """
+        Enqueue a new `BackgroundJob`.
+
+        This method is a wrapper of `Job.enqueue()` using `handle()` as function callback. See its documentation for
+        parameters. Note that specifying a custom `name` is not supported, as a `ScheduledJob` is identified by the job
+        class `__name__` automatically.
+        """
+        kwargs.pop('name', None)
+        return super().enqueue(name=cls.__name__, *args, **kwargs)
+
+    @classmethod
+    @advisory_lock(ADVISORY_LOCK_KEYS['job-schedules'])
+    def schedule(cls, instance, interval=None, *args, **kwargs):
+        """
+        Schedule a `ScheduledJob`.
+
+        This method adds a new `ScheduledJob` to the job queue. If the job schedule identified by its `instance` and
+        name is already active, the existing job will be updated if needed. However, this doesn't forbid running
+        additional jobs using the `enqueue()` method, e.g. to schedule an immediate synchronization job in addition to
+        periodic synchronization scheduled by this method.
+
+        For additional parameters see `enqueue()`.
+
+        Args:
+            instance: The NetBox object to which this `ScheduledJob` pertains
+            interval: Recurrence interval (in minutes)
+        """
+        job = cls.get_jobs(instance).filter(status__in=cls.ENQUEUED_STATUS).first()
+        if job:
+            # If the job parameters haven't changed, don't schedule a new job and keep the current schedule. Otherwise,
+            # delete the existing job and schedule a new job instead.
+            if job.interval == interval:
+                return job
+            job.delete()
+
+        return cls.enqueue(instance=instance, interval=interval, *args, **kwargs)
+
+
+class SystemJob(ScheduledJob):
```

Comment:
Would it remove the need for this class if we allowed a Job to be enqueued without defining an object? I don't actually think it was intended to be required: The Job model currently requires an `object_type` but not an `object_id`. I'm open to removing that requirement if it simplifies this effort.

Reply:
Yep. The sole purpose of this class is to circumvent the required `instance` parameter of `Job.enqueue()` and adding the `setup()` method. I'll add a commit to make the `instance` parameter optional and thus removing the need for this class.